requests-for-comments/2023-03-01-Temporal-RFC.md

```diff
@@ -0,0 +1,239 @@
+
+[https://github.com/posthog/energize](https://github.com/posthog/energize)
+
+Note: that this will be folded into the Posthog/Posthog repo. For now this is separate for example purposes.
+
+The specific workflow for S3 exporting can be found [here](https://github.com/PostHog/energize/blob/main/src/energize/workflows/exports/s3.py).
+
+# Temporal Cloud
+
+Why ☁️?
+
+There is one critical piece of the Temporal topology that causes loss if it goes down: The scheduler. As long as the scheduler is up. When you schedule a job - it will be completed. If the scheduler is down for an extended period of time there is a chance that the requesting service would drop the intent and that workflow would never be run. As long as the scheduler is up and we are with the tunable TTL of the workflow, as soon as a worker comes back online we will process that workflow.
+
+Luckily [Temporal.io](http://Temporal.io) has a hosted version of their scheduler where we can just submit our workflows and connect our workers to and the rest should be relatively smooth sailing. We only have to worry about writing the workflows themselves and keeping the workers happy.
+
+# Ownership
+
+This is why I’ve wanted to get you all here. Here is how ownership here would most likely shake out:
+
+**Infrastructure**
+
+- K8s Deployment of the workers that execute Workflows and Actions
+- CD of these workers
+
+**Billing**
+
+- Workflows and Actions written to be executed on Temporal workers
+
+**Pipeline**
+
+- Same, Workflows and actions written and scheduled to be executed on Temporal workers
+
+**Temporal cloud**
+
+- The scheduler itself. As long as the workflows are scheduled, they will be executed at some point in the future as long as a worker starts back up and the TTL does not expire the job.
+
+# Phases of work
+
+**Phase 1: Manual mode -** We disable historical exports in the app interface. This is the buggiest feature that larger high value customers are complaining about. This transitions into a manual mode where if a customer wants a historical export they request that we run one for them and we do that manually for them using a Temporal workflow. This allows us to be iterative and to start gathering operational knowledge around Temporal.
+
+**Phase 2: Build the Platform -** We stand up Temporal cloud as our scheduler and add worker deployments to K8s. This will build the foundation for migrating tasks to Temporal and unblock those that are interested in migrating their workflows to Temporal. Manual exports will run on the Temporal workers that are deployed on the K8s cluster. Temporal worker code is in the PostHog/PostHog monorepo
+
+**Phase 3: API-ification -** We rewire the historical exports interfaces to schedule Temporal workflows. The api is already there (built for the manual backfills). The only difference now is that we are scheduling workflows from the PostHog application through the user interface
```

Comment:
I'd lvoe a bit more detail around this point. Does this mean each plugin will have code in two places (real time and historical), written in potentially two different languages? Will we port all the plugins?

Reply:
Because temporal supports mixing tasks written in different languages within a workflow, I hope we'll be able to shim existing apps into temporal tasks (both for imports and exports)
While there's value in making a few most-used integrations more "temporal-native", I'd like to address the long tail with such a shim.